Sources/GRPC/ConnectionPool/ConnectionPool.swift

```diff
@@ -0,0 +1,660 @@
+      ()
+    }
+  }
+
+  func connectionIsQuiescing(_ manager: ConnectionManager) {
+    self.eventLoop.assertInEventLoop()
+    guard let removed = self.connections.removeValue(forKey: manager.id) else {
+      return
+    }
+
+    // Drop any delegates. We're no longer interested in these events.
+    removed.manager.sync.connectivityDelegate = nil
+    removed.manager.sync.http2Delegate = nil
+
+    // Replace the connection with a new idle one.
+    self.addConnectionToPool()
+
+    // Since we're removing this connection from the pool, the pool manager can ignore any streams
+    // reserved against this connection.
+    //
+    // Note: we don't need to adjust the number of available streams as the number of connections
+    // hasn't changed.
+    self.streamLender.returnStreams(removed.reservedStreams, to: self)
+  }
+
+  /// A connection is now in the ready state, it can provide streams to waiters.
+  private func connectionAvailable(_ id: ConnectionManagerID) {
+    self.eventLoop.assertInEventLoop()
+    self.connections[id]?.available(maxConcurrentStreams: self.maxConcurrentStreams)
+
+    // A connection becoming ready is usually followed by an update to max concurrent streams.
+    // Instead of servicing waiters right now we'll do this in a jiffy, i.e. once we receive the
+    // update to max concurrent streams. This allows us to avoid succeeding too many waiters if max
+    // concurrent streams is lower than we previously assumed.
+    self.eventLoop.execute {
+      self.tryServiceWaiters()
+    }
+  }
+
+  /// A connection has become unavailable.
+  private func connectionUnavailable(_ id: ConnectionManagerID) {
+    self.eventLoop.assertInEventLoop()
+    // The connection is no longer available: any streams which haven't been closed will be counted
+    // as a dropped reservation, we need to tell the pool manager about them.
+    if let droppedReservations = self.connections[id]?.unavailable(), droppedReservations > 0 {
+      self.streamLender.returnStreams(droppedReservations, to: self)
+    }
+  }
+}
+
+extension ConnectionPool: ConnectionManagerHTTP2Delegate {
+  internal func streamClosed(_ manager: ConnectionManager) {
+    self.eventLoop.assertInEventLoop()
+
+    // Return the stream the connection and to the pool manager.
+    self.connections[manager.id]?.returnStream()
+    self.streamLender.returnStreams(1, to: self)
+
+    // A stream was returned: we may be able to service a waiter now.
+    self.tryServiceWaiters()
+  }
+
+  internal func receivedSettingsMaxConcurrentStreams(
+    _ manager: ConnectionManager,
+    maxConcurrentStreams: Int
+  ) {
+    self.eventLoop.assertInEventLoop()
+    self.connections[manager.id]?.updateMaxConcurrentStreams(maxConcurrentStreams)
+
+    if maxConcurrentStreams != self.maxConcurrentStreams {
+      self.maxConcurrentStreams = maxConcurrentStreams
+      let maxAvailable = maxConcurrentStreams * self.connections.count
```

Comment:
Is this an abstraction that we are allowed to make? Can we assume that if we are allowed to send more streams on one connection, this will apply to all connections?

Reply:
It would be better if we didn't.